### PR TITLE
Use `delay_span_bug` for error cases when checking `AnonConst` parent

### LINKED ISSUE #60650: ICE with const generics in array definition: unexpected const parent path def
The following code snippet ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5096f7a9c82f038e5ef0eb184e4839f3)):
``` rust
#![feature(const_generics)]

struct Foo<const NUM_BYTES: usize>(pub [u8; NUM_BYTES]);

fn main() {
    let foo = Foo::<3>([1, 2, 3]);
}
```

ICEs with the following error:
```
   Compiling playground v0.0.1 (/playground)
warning: the feature `const_generics` is incomplete and may cause the compiler to crash
 --> src/main.rs:1:12
  |
1 | #![feature(const_generics)]
  |            ^^^^^^^^^^^^^^

error: internal compiler error: src/librustc_typeck/collect.rs:1407: unexpected const parent path def Def(Ctor(Struct, Fn), DefId(0/1:9 ~ playground[c276]::Foo[0]::{{constructor}}[0]))

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:637:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::util::bug::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: rustc::ty::context::tls::with_context_opt
  11: rustc::ty::context::tls::with_opt
  12: rustc::util::bug::opt_span_bug_fmt
  13: rustc::util::bug::bug_fmt
  14: rustc_typeck::collect::checked_type_of
  15: rustc_typeck::collect::type_of
  16: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute
  17: rustc::dep_graph::graph::DepGraph::with_task_impl
  18: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  19: rustc::ty::instance::Instance::resolve
  20: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_const
  21: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  22: rustc::ty::fold::TypeFoldable::fold_with
  23: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable for &rustc::ty::TyS>::super_fold_with
  24: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  25: rustc::traits::project::normalize
  26: rustc_typeck::check::FnCtxt::instantiate_value_path
  27: rustc_typeck::check::FnCtxt::check_expr_kind
  28: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_needs
  29: rustc_typeck::check::callee::<impl rustc_typeck::check::FnCtxt>::check_call
  30: rustc_typeck::check::FnCtxt::check_expr_kind
  31: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_needs
  32: rustc_typeck::check::FnCtxt::check_decl_initializer
  33: rustc_typeck::check::FnCtxt::check_decl_local
  34: rustc_typeck::check::FnCtxt::check_stmt
  35: rustc_typeck::check::FnCtxt::check_block_with_expected
  36: rustc_typeck::check::FnCtxt::check_expr_kind
  37: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_needs
  38: rustc_typeck::check::FnCtxt::check_return_expr
  39: rustc_typeck::check::check_fn
  40: rustc::ty::context::GlobalCtxt::enter_local
  41: rustc_typeck::check::typeck_tables_of
  42: rustc::ty::query::__query_compute::typeck_tables_of
  43: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute
  44: rustc::dep_graph::graph::DepGraph::with_task_impl
  45: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  46: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  47: rustc_typeck::check::typeck_item_bodies
  48: rustc::ty::query::__query_compute::typeck_item_bodies
  49: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_item_bodies>::compute
  50: rustc::dep_graph::graph::DepGraph::with_task_impl
  51: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  52: rustc::util::common::time
  53: rustc_typeck::check_crate
  54: rustc_interface::passes::analysis
  55: rustc::ty::query::__query_compute::analysis
  56: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  57: rustc::dep_graph::graph::DepGraph::with_task_impl
  58: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  59: rustc::ty::context::tls::enter_global
  60: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  61: rustc_interface::passes::create_global_ctxt::{{closure}}
  62: rustc_interface::passes::BoxedGlobalCtxt::enter
  63: rustc_interface::interface::run_compiler_in_existing_thread_pool
  64: std::thread::local::LocalKey<T>::with
  65: scoped_tls::ScopedKey<T>::set
query stack during panic:
#0 [type_of] processing `main::{{constant}}#0`
#1 [typeck_tables_of] processing `main`
#2 [typeck_item_bodies] type-checking all item bodies
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.36.0-nightly (cfdc84a00 2019-05-07) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin
```

### LINKED ISSUE #60704: Stable 1.34.0 compiler panic unexpected const parent in type_of_def_id()
Following code sample causes a compiler panic. `<32>` should be `<u32>`.

```
use std::convert::TryInto;

fn main() {
    let x: u32 = 5i32.try_into::<32>().unwrap();
}
```

```
error: internal compiler error: src/librustc_typeck/collect.rs:1320: unexpected const parent in type_of_def_id(): Expr(expr(2289: entry.portfolio_id.try_into::<>()))

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:620:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.34.0 (91856ed52 2019-04-10) running on x86_64-unknown-linux-gnu

note: compiler flags: -C opt-level=3 -C debuginfo=2 -C linker=external/ts_cc/gcc5/bin/x86_64-unknown-linux-gnu-gcc -C link-args=-lstdc++ -lm -Wl,-no-as-needed -Wl,-z,relro,-z,now -pass-exit-codes -Wl,-fuse-ld=gold -L/usr/lib/x86_64-linux-gnu -L/lib/x86_64-linux-gnu -Wl,--gc-sections -static-libgcc --crate-type rlib
```

### PR DESCRIPTION
**Pull request by varkor**
Fixes #60704.
Fixes #60650.

### CODE REVIEW

File: src/test/ui/const-generics/cannot-infer-type-for-const-param.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#![feature(const_generics)]
+//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
+
+// We should probably be able to infer the types here. However, this test is checking that we don't
+// get an ICE in this case. It may be modified later to not be an error.
+
+struct Foo<const NUM_BYTES: usize>(pub [u8; NUM_BYTES]);
+
+fn main() {
+    let _ = Foo::<3>([1, 2, 3]); //~ ERROR type annotations needed
```
Comment (matthewjasper) [inline src/test/ui/const-generics/cannot-infer-type-for-const-param.rs]: What happens if this is replaced with `3usize`?

Comment (varkor) [reply src/test/ui/const-generics/cannot-infer-type-for-const-param.rs]: Unfortunately, it ICEs:
```
src/librustc_codegen_llvm/context.rs:862: failed to get layout for `[type error]`: the type `[type error]` has an unknown layout
```
I think this is probably a result of the [existing issues](https://github.com/rust-lang/rust/issues/60619) with const generics and arrays. I'll add the updated test as a new issue.

**Comment by varkor** [comment]
Beta-nominating, because this fixes an ICE on stable/beta that seems like it could happen not uncommonly from a typo.

**Comment by pnkfelix** [comment]
accepted for beta-backport

**Comment by pietroalbini** [comment]
@varkor the `ui/const-generics/cannot-infer-type-for-const-param.rs` test still ICEs when backported to beta. Can you prepare a PR backporting the changes to beta yourself before Monday?

**Comment by varkor** [comment]
@pietroalbini: I will try, but it might be okay just to leave that test out entirely, as it only occurs under the feature flag, which can't be enabled on beta.

**Comment by pietroalbini** [comment]
@rust-lang/compiler what do you want to do then? Leave out the test?

**Comment by oli-obk** [comment]
yea, just leave out the test

**Comment by pnkfelix** [comment]
for the record: I agree with @oli-obk. Leave out the test.

**Pull request merged**